### PR TITLE
KAFKA-17248: Add reporter for adding thread metrics to telemetry pipeline and a test [2/N]

### PR DESCRIPTION
**Pull request by bbejeck**
This PR adds a `Reporter` instance that will add streams thread metrics to the telemetry pipeline.

For testing, the PR adds a unit test.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            log.debug("Registering metric {}", metric.metricName());
+            consumer.registerMetricForSubscription(metric);
+        }
+    }
+
+    private boolean tagMatchStreamOrStateUpdaterThreadId(final KafkaMetric metric) {
+        final Map<String, String> tags = metric.metricName().tags();
+        final boolean shouldInclude = tags.containsKey(THREAD_ID_TAG) && (tags.get(THREAD_ID_TAG).equals(threadId) || tags.get(THREAD_ID_TAG).equals(stateUpdaterThreadId));
+        if (!shouldInclude) {
+            log.warn("Rejecting metric {}", metric.metricName());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java]: This should not be WARN (maybe DEBUG or even better TRACE)?

Given that we plan to pass in the full KS metric object in here, we do expect that many metrics will be rejected (ie, all the client-level metrics) so it's nothing we need to "warn" users about.

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        return shouldInclude;
+    }
+
+    @Override
+    public void metricRemoval(final KafkaMetric metric) {
+        if (tagMatchStreamOrStateUpdaterThreadId(metric)) {
+            log.debug("Unregistering metric {}", metric.metricName());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java:72]: This one TRACE too maybe?

**Review by mjsax — APPROVED** [review]
Two minor comments. Feel free to merge after addressed.

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public void init(final List<KafkaMetric> metrics) {
+        metrics.forEach(this::metricChange);
+    }
+
+    @Override
+    public void metricChange(final KafkaMetric metric) {
+        if (tagMatchStreamOrStateUpdaterThreadId(metric)) {
+            log.debug("Registering metric {}", metric.metricName());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java:55]: Change to `trace`?

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java:72]: as discussed - keeping at debug

File: streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java (hunk 3 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/internals/metrics/StreamsThreadMetricsDelegatingReporter.java:55]: as discussed keeping at debug

**Comment by bbejeck** [comment]
Merged #17376 into trunk

**Pull request merged**